Fix DataToolParameters accepting input with arbitrary datatype in tests

bernt-matthias opened the pull request:
In the first place this PR added a test case (`test/functional/tools/bam_input_for_sam.xml`) showing that in tests (e.g. with `planemo test`) data parameters of tools accept input of arbitrary datatypes. The test shows this for a `sam` input which gets `bam` and `bed` in the two tests (the `bam` case might be interpreted as a case of a missing converter as @mvdbeek pointed out below).

Note that in the UI this works as expected, i.e. only sam is accepted. 

The main fix is in lines:

https://github.com/galaxyproject/galaxy/blob/570686ee62937aaf53ce5a12ea959260a23a875f/lib/galaxy/tools/parameters/basic.py#L1961-L1966

Previously the loop containing this code checked only if an implicit conversion is possible/necessary for one/more inputs. Now an exception is raised if the datasets do not match the datatype required by the parameter.
An additional necessary change is that `dataset_matcher.hda_match` now uses `ensure_visible=False`. The consequence is that also hidden datasets are considered. This was necessary since otherwise collection input does not get a match if the contained data sets are hidden (makes me wonder if implicit conversion ever worked for this case .. edit: seems so, but I don't understand why). _Not sure about other possible side effects_. 

This required a quite few changes to tests where datatypes of inputs were not specified or the wrong datatype was shown. While fixing I realized that at different places `ext`, `extension`, and `file_type` are used .. would be nice if this would become consistent...

TODO

- [ ] the same in necessary for `DataCollectionToolParameter`, somewhere here https://github.com/galaxyproject/galaxy/blob/433874d112d48a11339d3fbe2a5e65531b9e4cd6/lib/galaxy/tools/parameters/basic.py#L2168
  - [ ] add a test
  - [ ] add fix
- [ ] adapt Exception text. Since a mismatch can alslo happen if an options filter kicks out datasets with wrong / missing dbkey

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
add a test case that shows that sam input accepts bam

which it should not

bernt-matthias committed:
check for datatype mismatch

bernt-matthias committed:
fix one api test

bernt-matthias committed:
fix unit tests

bernt-matthias committed:
fix 2 api tests

bernt-matthias committed:
also match invisible datasets

otherwise if a tool is mapped over a collection where the
included datasets are hidden does not work.
I guess also implicit conversion did not work so far for this case.

Not sure about possible side effects. I tested manually
that the tool form does not include hidden datasets

bernt-matthias committed:
fix more api tests

and potentially one selenium test

bernt-matthias committed:
fix framework tests

bernt-matthias committed:
fix selenium test

bernt-matthias committed:
try to fix last failing api and selenium test

bernt-matthias committed:
raise a more helpful error

bernt-matthias committed:
undo set format for test cat tool

seems to fix api/selenium problems

bernt-matthias committed:
fix test assumption for bam4sam.xml

bernt-matthias committed:
add test for collections

bernt-matthias committed:
fixes for test tool

output does not matter anyway because of expect_failure

bernt-matthias committed:
first attempt to fix for collection inputs

bernt-matthias committed:
fix functional test

mvdbeek commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -2166,6 +2169,7 @@ def match_multirun_collections(self, trans, history, dataset_collection_matcher)
                 yield history_dataset_collection, match.implicit_conversion
 
     def from_json(self, value, trans, other_values=None):
+        log.error(f"from_json value {value} trans {trans} other_values {other_values}")
```
```suggestion
```

mvdbeek commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -2207,6 +2211,25 @@ def from_json(self, value, trans, other_values=None):
             # TODO: Handle error states, implement error states ...
+
+        # prepare dataset/collection matching
+        values = util.listify(rval)
```
`rval` should always be a single item, no need to call listify and treat it like a list

bernt-matthias replied:
OK 22d46f62c5

mvdbeek commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -2207,6 +2211,25 @@ def from_json(self, value, trans, other_values=None):
+        dataset_collection_matcher = dataset_matcher_factory.dataset_collection_matcher(dataset_matcher)
+
+        for v in values:
+            log.error(f"from_json v {v}")
```
```suggestion
```

mvdbeek commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -2207,6 +2211,25 @@ def from_json(self, value, trans, other_values=None):
+                # not sure if implicit conversion needs to be handled here
+                # if implicit_conversion:
+                #     v.implicit_conversion = True
+            else:
```
```suggestion
            match = dataset_collection_matcher.hdca_match(v)
            if not match:

```

mvdbeek replied:
The only place that looks at this attribute is https://github.com/mvdbeek/galaxy/blob/b26e8ae3f581bedb95b5b421c4879574116bff87/lib/galaxy/tools/parameters/basic.py#L88, and that is preceded by `isinstance(v, HistoryDatasetAssociation)`.

mvdbeek commented on lib/galaxy_test/base/populators.py:
```diff
@@ -1674,19 +1675,25 @@ def _create_collection(self, payload: dict) -> Response:
+        if filename is not None:
+            return open(filename, mode)
+        else:
+            return StringIO(test_dict["value"])
```
This is an optional suggestion totally biased by having done this in https://github.com/galaxyproject/galaxy/pull/13110/files, but I like the distinction CWL makes here. If `content` is provided it is used as a file literal. So instead of deciding this based on whether value is in the test data dir, you could restore the original version and use something like
```python
            input_type = value["type"]
            if input_type == "File":
                if 'content' in value:
                    content = value.get('content')
                else:
                    content = open_test_data(value)
```
on line 1738, together with changing `value` to `content` in the test_data

```suggestion
        return open(filename, mode)
```

jmchilton replied:
Yeah - we cannot just rework what value here means I think given that we want to work on how value is resolved.

mvdbeek commented on lib/galaxy_test/base/workflow_fixtures.py:
```diff
@@ -199,12 +199,18 @@
-    samp1\t30.0
-    samp2\t40.0
+  text_input1:
+      value: |
```
```suggestion
      content: |
```

mvdbeek commented on lib/galaxy_test/base/workflow_fixtures.py:
```diff
@@ -199,12 +199,18 @@
+      type: File
+      file_type: tabular
+  text_input2:
+      value: |
```
```suggestion
      content: |
```

mvdbeek reviewed (approved):
Very nice, I like it, just some minor edits needed.

mvdbeek commented on lib/galaxy/tools/parameters/basic.py:
```diff
@@ -1958,16 +1958,19 @@ def from_json(self, value, trans, other_values=None):
+                    if match:
+                        if match.implicit_conversion:
+                            v.implicit_conversion = True
+                    else:
+                        raise ParameterValueError(f"data set ({v.name}) with invalid datatype ({v.ext}) supplied to input dataset parameter (expecting {dataset_matcher.param.extensions})", self.name)
```
Maybe we should allow `data` as an extension here for the CWL work that doesn't really use extension and for remote data that doesn't have an extension yet? I guess we've benefited this far from this kind of "loophole".

bernt-matthias replied:
I would have guessed since essentially this function is called https://github.com/galaxyproject/galaxy/blob/9bbc7e7206035cf9a796264f578c5a172b4f8309/lib/galaxy/datatypes/data.py#L841 or is `isinstance` working the other way round? 

If I'm wrong I guess checking for `data` should be done in the matcher, or?

bernt-matthias committed:
Apply suggestions from code review

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
fix 1 test

by using content instead of value

bernt-matthias committed:
add functional tool test for data parameter options filter

bernt-matthias committed:
fix one more api test

bernt-matthias committed:
do not listify rval

as suggested here https://github.com/galaxyproject/galaxy/pull/12073#discussion_r785038851

and distinguish HDCA and DCE

bernt-matthias committed:
try suggestion

bernt-matthias committed:
Revert "try suggestion"

This reverts commit a1625e28817aae23abd0cb461f9591b97eb7c2e9.
